### PR TITLE
parser: implement Restore for ColumnOption, FieldType and ColumnDef

### PR DESCRIPTION
**Pull request by feloxx**
1. implement Restore FieldType
2. implement Restore ColumnDef ColumnOption
3. Change OpCode io => ctx

### CODE REVIEW

**Comment by leoppro** [comment]
PTAL @kennytm

File: ast/ddl_test.go [hunk 1]
```diff
@@ -164,6 +164,81 @@ func (ts *testDDLSuite) TestDDLConstraintRestore(c *C) {
...
+		{"id timestamp", "`id` TIMESTAMP"},
+		{"id datetime(2)", "`id` DATETIME(2)"},
+		{"id date", "`id` DATE"},
+		{"id year", "`id` YEAR"},
+		{"id INT", "`id` INT"},
+		{"id INT NULL", "`id` INT NULL"},
+
+		{"id int(11) PRIMARY KEY", "`id` INT(11) PRIMARY KEY"},
+		{"id int(11) NOT NULL", "`id` INT(11) NOT NULL"},
+		{"id INT(11) NULL", "`id` INT(11) NULL"},
```
Comment (leoppro) [inline ast/ddl_test.go]: add test for enum and set

Comment (leoppro) [reply ast/ddl_test.go]: and I think the test cases here should cover all types.

File: types/field_type.go [hunk 2]
```diff
@@ -195,6 +195,61 @@ func (ft *FieldType) String() string {
...
+	if mysql.HasBinaryFlag(ft.Flag) && ft.Tp != mysql.TypeString {
+		ctx.WriteKeyWord(" BINARY")
+	}
+
+	if IsTypeChar(ft.Tp) || IsTypeBlob(ft.Tp) {
+		if ft.Charset != "" && ft.Charset != charset.CharsetBin {
+			ctx.WriteKeyWord(" CHARACTER SET " + ft.Charset)
+		}
+		if ft.Collate != "" && ft.Collate != charset.CharsetBin {
+			ctx.WriteKeyWord(" COLLATE ")
```
Comment (kennytm) [inline types/field_type.go]: The COLLATE modifier can appear independently without the CHARACTER SET modifier:

```sql
mysql> create table a (a varchar(50) collate utf8mb4_czech_ci);
Query OK, 0 rows affected (0.10 sec)
```

Could you verify if it works correctly?

File: ast/ddl_test.go [hunk 3]
```diff
@@ -164,6 +164,88 @@ func (ts *testDDLSuite) TestDDLConstraintRestore(c *C) {
...
+		{"id tinyint(4)", "`id` TINYINT(4)"},
+		{"id smallint(6)", "`id` SMALLINT(6)"},
+		{"id mediumint(9)", "`id` MEDIUMINT(9)"},
+		{"id integer(11)", "`id` INT(11)"},
+		{"id bigint(20)", "`id` BIGINT(20)"},
+		{"id DATE", "`id` DATE"},
+		{"id DATETIME", "`id` DATETIME"},
+		{"id DECIMAL(4,2)", "`id` DECIMAL(4,2)"},
+		{"id char(1)", "`id` CHAR(1)"},
+		{"id varchar(10) BINARY", "`id` VARCHAR(10) BINARY"},
```
[the remark this answers is not in the document]
Comment (feloxx) [reply ast/ddl_test.go]: thank you

File: types/field_type.go (hunk 2 above)
Comment (feloxx) [reply types/field_type.go]: thank you

**Pull request merged**